File: go/audit_log.go
// audit_log.go
//
// Ports CircleAI.Core.Auditing.ICircleAIAuditLog + CircleAIAuditEntry +
// CircleAIAuditQuery (ICircleAIAuditLog.cs), CircleAI.Core.Auditing.LoggerAuditLog
// (LoggerAuditLog.cs), CircleAI.Core.Auditing.NoopAuditLog (NoopAuditLog.cs), and
// the ambient accessor CircleAI.Core.Auditing.CircleAIAuditing (CircleAIAuditing.cs).
//
// The audit surface records every state-changing SDK operation. Record MUST
// NOT fail the caller — implementations fail open. C# QueryAsync yields an
// IAsyncEnumerable; the Go port returns a slice (the stream carries only
// entries, no incremental semantics to preserve).

package circleai

import (
	"context"
	"fmt"
	"sync"
	"time"
)

// CircleAIAuditEntry is one immutable audit entry. Ports CircleAIAuditEntry.
type CircleAIAuditEntry struct {
	At               time.Time // UTC timestamp of the action (required)
	Component        string    // canonical component name (required)
	Operation        string    // logical operation name (required)
	Outcome          string    // outcome, one of the Outcomes constants (required)
	TenantID         string    // tenant id when multi-tenant; empty otherwise
	UhidIdentityID   string    // UHID when user-scoped; empty otherwise
	CorrelationID    string    // optional correlation id (session/request)
	DurationMs       float64   // operation duration in milliseconds
	ErrorType        string    // CLR/Go error type when Outcome != success
	ErrorCode        string    // implementation-supplied error code
	PayloadSha256Hex string    // hash of any sensitive payload; never the payload
}

// CircleAIAuditQuery filters QueryAsync. Ports CircleAIAuditQuery. A zero value
// has MaxItems == 0; use NewCircleAIAuditQuery for the C# default of 1000.
type CircleAIAuditQuery struct {
	FromUTC        *time.Time // inclusive lower bound on At
	ToUTC          *time.Time // inclusive upper bound on At
	Component      string     // restrict to a single component
	TenantID       string     // restrict to a single tenant
	UhidIdentityID string     // restrict to a single UHID
	Outcome        string     // restrict to a single outcome
	MaxItems       int        // max entries to return
}

// NewCircleAIAuditQuery returns a query with the C# default MaxItems (1000).
func NewCircleAIAuditQuery() CircleAIAuditQuery {
	return CircleAIAuditQuery{MaxItems: 1000}
}

// ICircleAIAuditLog is the audit sink contract. Ports ICircleAIAuditLog.
type ICircleAIAuditLog interface {
	// Record appends an entry. MUST NOT return an error that would abort the
	// caller — implementations fail open. The error return exists only for
	// implementations that choose to surface a non-fatal note.
	Record(ctx context.Context, entry CircleAIAuditEntry) error

	// Query returns historical entries matching the filter.
	Query(ctx context.Context, query CircleAIAuditQuery) ([]CircleAIAuditEntry, error)
}

// NoopAuditLog silently discards every entry and returns no query results.
// Ports CircleAI.Core.Auditing.NoopAuditLog.
type NoopAuditLog struct{}

// NoopAuditLogInstance is the shared singleton. Mirrors NoopAuditLog.Instance.
var NoopAuditLogInstance = NoopAuditLog{}

// Record discards the entry.
func (NoopAuditLog) Record(context.Context, CircleAIAuditEntry) error { return nil }

// Query returns an empty result.
func (NoopAuditLog) Query(context.Context, CircleAIAuditQuery) ([]CircleAIAuditEntry, error) {
	return nil, nil
}

// AuditLogSink receives a formatted structured audit line. This is the
// injection point that replaces the C# ILogger in LoggerAuditLog.
type AuditLogSink func(message string)

// LoggerAuditLog writes structured entries to an injected sink at
// information level. Query always returns empty — reading back from a log
// pipeline is not possible at the SDK layer. Ports LoggerAuditLog.
type LoggerAuditLog struct {
	sink AuditLogSink
}

// NewLoggerAuditLog builds a logger-backed audit log over the given sink
// (required). Mirrors the C# ctor's null guard.
func NewLoggerAuditLog(sink AuditLogSink) (*LoggerAuditLog, error) {
	if sink == nil {
		return nil, fmt.Errorf("sink is required")
	}
	return &LoggerAuditLog{sink: sink}, nil
}

// Record formats the entry as a structured line and hands it to the sink. The
// template mirrors the C# LogInformation call, with "-" for empty fields.
func (l *LoggerAuditLog) Record(_ context.Context, entry CircleAIAuditEntry) error {
	l.sink(fmt.Sprintf(
		"CircleAI audit %s.%s %s tenant=%s uhid=%s corr=%s duration_ms=%g error=%s(%s) payload_sha256=%s at=%s",
		entry.Component, entry.Operation, entry.Outcome,
		orDash(entry.TenantID), orDash(entry.UhidIdentityID), orDash(entry.CorrelationID),
		entry.DurationMs, orDash(entry.ErrorType), orDash(entry.ErrorCode),
		orDash(entry.PayloadSha256Hex), entry.At.UTC().Format(time.RFC3339Nano)))
	return nil
}

// Query always returns empty.
func (l *LoggerAuditLog) Query(context.Context, CircleAIAuditQuery) ([]CircleAIAuditEntry, error) {
	return nil, nil
}

func orDash(s string) string {
	if s == "" {
		return "-"
	}
	return s
}

// ── CircleAIAuditing ambient accessor ─────────────────────────────────────

// circleAIAuditing is the process-wide ambient audit sink. Ports the static
// CircleAI.Core.Auditing.CircleAIAuditing. Guarded for concurrent Set/Get.
var circleAIAuditing = struct {
	mu      sync.RWMutex
	current ICircleAIAuditLog
}{current: NoopAuditLogInstance}

// AuditingDefault returns the current ambient audit sink (default NoopAuditLog).
// Ports CircleAIAuditing.Default.
func AuditingDefault() ICircleAIAuditLog {
	circleAIAuditing.mu.RLock()
	defer circleAIAuditing.mu.RUnlock()
	return circleAIAuditing.current
}

// SetAuditingDefault replaces the ambient audit sink (required, non-nil).
// Ports CircleAIAuditing.SetDefault.
func SetAuditingDefault(audit ICircleAIAuditLog) error {
	if audit == nil {
		return fmt.Errorf("audit sink is required")
	}
	circleAIAuditing.mu.Lock()
	defer circleAIAuditing.mu.Unlock()
	circleAIAuditing.current = audit
	return nil
}

// ResetAuditingToNoop restores the ambient sink to NoopAuditLog. Test helper.
// Ports CircleAIAuditing.ResetToNoop.
func ResetAuditingToNoop() {
	circleAIAuditing.mu.Lock()
	defer circleAIAuditing.mu.Unlock()
	circleAIAuditing.current = NoopAuditLogInstance
}

var (
	_ ICircleAIAuditLog = NoopAuditLog{}
	_ ICircleAIAuditLog = (*LoggerAuditLog)(nil)
)

File: python/src/circle_ai/core/audit_log.py
# core/audit_log.py
#
# Port of CircleAI.Core.Auditing:
#   • ICircleAIAuditLog       — tamper-aware audit surface
#   • CircleAIAuditEntry      — immutable audit entry record
#   • CircleAIAuditQuery      — query filter
#   • NoopAuditLog            — default: silently drops entries
#   • LoggerAuditLog          — writes structured entries to a logging.Logger
#   • CircleAIAuditing        — process-wide ambient access point
#
# The C# QueryAsync is an IAsyncEnumerable; the Python contract is an async
# generator (``async def ... yield``) so ``async for`` works identically.

from __future__ import annotations

import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from typing import AsyncIterator, Optional


# ─────────────────────────────────────────────────────────────────────────────
# CircleAIAuditEntry — immutable audit entry emitted by the SDK.
# ─────────────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class CircleAIAuditEntry:
    """An immutable audit entry emitted by the CircleAI SDK.

    ``at``, ``component``, ``operation`` and ``outcome`` are required (they map
    to the C# ``required`` init-only properties); the rest are optional.
    """

    at: datetime
    component: str
    operation: str
    outcome: str
    tenant_id: Optional[str] = None
    uhid_identity_id: Optional[str] = None
    correlation_id: Optional[str] = None
    duration_ms: float = 0.0
    error_type: Optional[str] = None
    error_code: Optional[str] = None
    payload_sha256_hex: Optional[str] = None


# ─────────────────────────────────────────────────────────────────────────────
# CircleAIAuditQuery — query filter for ICircleAIAuditLog.query_async.
# ─────────────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class CircleAIAuditQuery:
    """Query filter for :meth:`ICircleAIAuditLog.query_async`."""

    from_utc: Optional[datetime] = None
    to_utc: Optional[datetime] = None
    component: Optional[str] = None
    tenant_id: Optional[str] = None
    uhid_identity_id: Optional[str] = None
    outcome: Optional[str] = None
    max_items: int = 1000


# ─────────────────────────────────────────────────────────────────────────────
# ICircleAIAuditLog — the audit contract.
# ─────────────────────────────────────────────────────────────────────────────


class ICircleAIAuditLog(ABC):
    """Tamper-aware audit surface for the CircleAI SDK.

    Every state-changing operation a component performs is auto-recorded here.
    Default registration is :class:`NoopAuditLog` — entries are silently
    dropped until a consumer wires :class:`LoggerAuditLog` or their own
    append-only sink.
    """

    @abstractmethod
    async def record_async(
        self, entry: CircleAIAuditEntry, ct: object = None
    ) -> None:
        """Record an audit entry. MUST NOT raise — the caller may be mid-operation
        and audit-log failure must never bring it down. Implementations should
        catch and log internally, failing open."""
        raise NotImplementedError

    @abstractmethod
    def query_async(
        self, query: CircleAIAuditQuery, ct: object = None
    ) -> AsyncIterator[CircleAIAuditEntry]:
        """Query historical entries — for compliance reporting, forensic
        investigation, debugging. Returns an async iterator."""
        raise NotImplementedError


# ─────────────────────────────────────────────────────────────────────────────
# NoopAuditLog — default sink; silently discards every entry.
# ─────────────────────────────────────────────────────────────────────────────


class NoopAuditLog(ICircleAIAuditLog):
    """Default :class:`ICircleAIAuditLog` — silently discards every entry and
    returns an empty query result."""

    _instance: Optional["NoopAuditLog"] = None

    @classmethod
    def instance(cls) -> "NoopAuditLog":
        """Shared singleton instance (mirrors C# ``NoopAuditLog.Instance``)."""
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    async def record_async(
        self, entry: CircleAIAuditEntry, ct: object = None
    ) -> None:
        return None

    async def query_async(
        self, query: CircleAIAuditQuery, ct: object = None
    ) -> AsyncIterator[CircleAIAuditEntry]:
        return
        yield  # pragma: no cover — makes this an async generator that yields nothing


# ─────────────────────────────────────────────────────────────────────────────
# LoggerAuditLog — writes structured entries to a logging.Logger.
# ─────────────────────────────────────────────────────────────────────────────


class LoggerAuditLog(ICircleAIAuditLog):
    """:class:`ICircleAIAuditLog` that writes structured entries to a
    :class:`logging.Logger` at ``INFO`` level.

    The :meth:`query_async` implementation always returns empty — reading back
    from a logger isn't possible at the SDK layer.
    """

    def __init__(self, logger: logging.Logger) -> None:
        if logger is None:
            raise ValueError("logger")
        self._logger = logger

    async def record_async(
        self, entry: CircleAIAuditEntry, ct: object = None
    ) -> None:
        if entry is None:
            raise ValueError("entry")
        self._logger.info(
            "CircleAI audit %s.%s %s tenant=%s uhid=%s corr=%s "
            "duration_ms=%s error=%s(%s) payload_sha256=%s at=%s",
            entry.component,
            entry.operation,
            entry.outcome,
            entry.tenant_id or "-",
            entry.uhid_identity_id or "-",
            entry.correlation_id or "-",
            entry.duration_ms,
            entry.error_type or "-",
            entry.error_code or "-",
            entry.payload_sha256_hex or "-",
            entry.at.isoformat(),
        )

    async def query_async(
        self, query: CircleAIAuditQuery, ct: object = None
    ) -> AsyncIterator[CircleAIAuditEntry]:
        return
        yield  # pragma: no cover


# ─────────────────────────────────────────────────────────────────────────────
# CircleAIAuditing — process-wide ambient access point.
# ─────────────────────────────────────────────────────────────────────────────


class CircleAIAuditing:
    """Process-wide ambient access point for the audit sink.

    Initial value is :meth:`NoopAuditLog.instance`. Hosts wire the real sink
    by calling :meth:`set_default` during startup.
    """

    _default: ICircleAIAuditLog = NoopAuditLog.instance()

    @classmethod
    def default(cls) -> ICircleAIAuditLog:
        """The current ambient audit sink."""
        return cls._default

    @classmethod
    def set_default(cls, audit: ICircleAIAuditLog) -> None:
        """Replace the ambient audit sink. Idempotent."""
        if audit is None:
            raise ValueError("audit")
        cls._default = audit

    @classmethod
    def reset_to_noop(cls) -> None:
        """Restore the default to :class:`NoopAuditLog`. Test-helper."""
        cls._default = NoopAuditLog.instance()
